server/graphql/mutations/orders.js

```diff
@@ -311,63 +309,71 @@ export function updateSubscription(remoteUser, args) {
     ]
   };
 
-  return models.Order.findOne(query)
-  .tap(order => {
-    if (!order) {
-      throw new Error("Subscription not found")
-    }
-    return Promise.resolve();
-  })
-  .tap(order => {
-    if (!remoteUser.isAdmin(order.FromCollectiveId)) {
-      throw new errors.Unauthorized({
-        message: "You don't have permission to update this subscription"
-      })
-    }
-    return Promise.resolve();
-  })
-  .tap(order => {
-    if (!order.Subscription.isActive) {
-      throw new Error('Subscription must be active to be updated');
-    }
-    return Promise.resolve();
-  })
-  .tap(order => {
+  let order = await models.Order.findOne(query);
 
-    let updatePromise;
-    let newPm;
-    
-    // TODO: Would be even better if we could charge you here directly
-    // before letting you proceed
-
-    // means it's an existing paymentMethod
-    if (paymentMethod.uuid && paymentMethod.uuid.length === 36) {
-      updatePromise = models.PaymentMethod.findOne({where: { uuid: paymentMethod.uuid}})
-        .then(pm => {
-          if (!pm){
-            throw new Error('Payment method not found with this uuid', paymentMethod.uuid);
-          }
-          newPm = pm;
-        })
-    } else {
-      // means it's a new paymentMethod
-      const newPMData = Object.assign(paymentMethod, { CollectiveId: order.FromCollectiveId })
-      updatePromise = models.PaymentMethod.createFromStripeSourceToken(newPMData)
-        .then(pm => newPm = pm)
-    }
+  if (!order) {
+    throw new Error("Subscription not found")
+  }
+  if (!remoteUser.isAdmin(order.FromCollectiveId)) {
+    throw new errors.Unauthorized({
+      message: "You don't have permission to update this subscription"
+    })
+  }
+  if (!order.Subscription.isActive) {
+    throw new Error('Subscription must be active to be updated');
+  }
 
-    // determine if this order was pastdue
-    if (order.Subscription.chargeRetryCount > 0) {
-      const updatedDates = getNextChargeAndPeriodStartDates('updated', order);
-      const chargeRetryCount = getChargeRetryCount('updated', order);
+  if (amount) {
+    order.Subscription.deactivate();
+
+    const newSubscriptionDataValues = Object.assign(pick(order.Subscription.dataValues, [
+      'interval',
+      'nextChargeDate',
+      'nextPeriodStart',
+      'stripeSubscriptionId',
+      'createdAt'
+    ]), {
+      amount: amount,
+      isActive: true,
+      updatedAt: new Date,
+      activatedAt: new Date,
+    });
+
+    const newSubscription = await models.Subscription.create(newSubscriptionDataValues);
+
+    order = await order.update({ totalAmount: amount, SubscriptionId: newSubscription.id });
```

Comment:
Any reason not to create a new `order` as well? Some tradeoffs of the top of my head:

- updating old order will create transactions with two different amounts for that order
- updating old order won't maintain history. User will only see the new subscription in their UI with no record of an old subscription with previous amount.
- One downside of new order is that it'll appear that we just activated that order. We could cheat by copying dates from previous order?

In general, `subscription` and `order` are 1:1 relationship and really don't even need to be separate tables (legacy reasons for their separation). My instinct is to lean towards maintaining that 1:1 relationship...

Thoughts? @clarete @xdamman

Reply:
@asood123 I can change that.

How would you approach testing? Having in mind that this we would now have new order and subscription ids. How do you check that they're properly "updated"? Retrieve the new order by collectiveId + userId?

Reply:
Yes, exactly. I'd search for all orders in the database by that user. Should return 2. One's subscription should be cancelled. Other is active and all the fields that were copied over should match.